Use more loose correction of the dependencies.

borissmidt opened the pull request:
This resolves the problem of having mixed crossbuild and non crossbuild projects. You can now configure any dependency with _2.11/_2.12/_2.13/_? and all will be accepted as a valid config.

prokod committed:
To update README.md

prokod committed:
To update credit

prokod committed:
To highlight version

prokod committed:
To remove SNAPSHOT state from project version

borissmidt committed:
Use more loose correction of the dependencies.

prokod commented:
Hi @borissmidt :)

First thing first - small clarification: The only real change (non cosmetic) happens in ResolutionStrategyConfigurer.groovy ?

borissmidt commented:
I think there are 2 i changed a regex and changed the fallback behavior
that if it find 0 alternatives that it just replaced the scala version with
the expected one.

But the regex change to replace the _? Doesn't seem to do much.

An alternative could be to just use an property to override the _?. For the
dependency tag. Many people already use a property anyway to mark there
scala version on all dependencies.

On Thu, Jul 23, 2020, 21:30 Noam Asor <notifications@github.com> wrote:

> Hi @borissmidt <https://github.com/borissmidt> :)
>
> First thing first - small clarification: The only real change (non
> cosmetic) happens in ResolutionStrategyConfigurer.groovy ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/prokod/gradle-crossbuild-scala/pull/88#issuecomment-663191351>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABNXZFW2UEHOSSJKSWLAOOTR5CFURANCNFSM4PGAVSVQ>
> .
>

borissmidt commented on src/main/groovy/com/github/prokod/gradle/crossbuild/ResolutionStrategyConfigurer.groovy:
```diff
@@ -223,7 +223,7 @@ class ResolutionStrategyConfigurer {
     private static void resolveQMarkDep(DependencyResolveDetails details, String replacementScalaVersion) {
         def requested = details.requested
-        def resolvedName = requested.name.replace('_?', "_$replacementScalaVersion")
+        def resolvedName = requested.name.replace('(_?|_2.11|2.12|2.13)', "_$replacementScalaVersion")
```
here its changed.

borissmidt commented on src/main/groovy/com/github/prokod/gradle/crossbuild/ResolutionStrategyConfigurer.groovy:
```diff
@@ -308,15 +308,19 @@ class ResolutionStrategyConfigurer {
-        def correctDependencyInsight = correctDependencies.head()
-        def correctDependency = correctDependencyInsight.dependency
+        if(correctDependencies.size() == 1) {
+            def correctDependencyInsight = correctDependencies.head()
```
and here

borissmidt commented on src/main/groovy/com/github/prokod/gradle/crossbuild/CrossBuildPlugin.groovy:
```diff
@@ -81,12 +81,12 @@ class CrossBuildPlugin implements Plugin<Project> {
 
             def sourceSet = getAndUpdateSourceSetFor(rb, extension.crossBuildSourceSets)
-
+            println("extension ${sourceSet} ${scalaVersionInsights}")
```
i should remove this.

borissmidt commented:
I cleaned up the PR

borissmidt commented on src/main/groovy/com/github/prokod/gradle/crossbuild/ResolutionStrategyConfigurer.groovy:
```diff
@@ -223,7 +223,7 @@ class ResolutionStrategyConfigurer {
     private static void resolveQMarkDep(DependencyResolveDetails details, String replacementScalaVersion) {
         def requested = details.requested
-        def resolvedName = requested.name.replace('_?', "_$replacementScalaVersion")
+        def resolvedName = requested.name.replaceFirst('_(\\?|2.11|2.12|2.13)', "_$replacementScalaVersion")
```
changed this to a regex

borissmidt commented on src/main/groovy/com/github/prokod/gradle/crossbuild/ResolutionStrategyConfigurer.groovy:
```diff
@@ -308,15 +308,19 @@ class ResolutionStrategyConfigurer {
+            // Assuming group is staying the same ...
+            offenderDetails.useTarget requested.group + ':' + correctDependency.name + ':' + correctDependency.version
+        } else {
+            offenderDetails.useTarget requested.group + ':' + requested.name.replaceFirst("(_2.11|_2.12|_2.13)",  "_${targetScalaVersion}") + ':' + requested.version
```
using a smarter correction if 0 alternatives are found.

prokod commented:
Hi @borissmidt :) Looks tidier now. Thx! Helps a lot with review process! my question will be how come we need to do replacement to actual _2.11, _2.12, _2.13 for the project to compile correctly ? you should either have _? dependencies where you want the plugin to expand it for you, based on the required cross compilation versions or, you as the user, should provide those deps in correct crosscompile configurations.

What am I missing ?

Could you provide output of `crossBuildResolvedConfigs` task on relevant modules please ?

borissmidt commented:
No we just use a separate dependency.gradle file. I just wonder why we need
an explicit _? Marker instead of just using _2.11 if you know your scala
version is 2.11.12 . Also in the example of a project with the play plugin
you can't combine it with your plugin so if it is defaulting then you don't
need any special dependency configs. I.e. the plugin just specifies the
version then and just uses the 2.11 2.12 2.13 as a marker. So other
projects don't see the difference.

On Fri, Jul 24, 2020, 22:22 Noam Asor <notifications@github.com> wrote:

> Hi @borissmidt <https://github.com/borissmidt> :) Looks tidier now. Thx!
> Helps a lot with review process! my question will be how come we need to do
> replacement to actual _2.11, _2.12, _2.13 for the project to compile
> correctly ? you should either have _? dependencies where you want the
> plugin to expand it for you, based on the required cross compilation
> versions or, you as the user, should provide those deps in correct
> crosscompile configurations.
>
> Are you hinting that you are maintaining explicit 3rd party scala based
> lib dependencies in build.gradle that are not aligned with default scala
> version as defined by explicit scala-library dependency?
>
> As an example:
>
> project A {
> apply plugin: 'com.github.prokod.gradle-crossbuild'
>
> dependencies {
> compile some-lib_2.12:1.2.3
> compile scala-lib:2.11.+
> }
> }
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/prokod/gradle-crossbuild-scala/pull/88#issuecomment-663718288>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABNXZFRLTOEOT5UFA46WFCDR5HURHANCNFSM4PGAVSVQ>
> .
>

borissmidt commented:
A better question is:

Would there be any case where you would want to have a _2.12 dependency in
a _2.13 build?

On Sat, Jul 25, 2020, 09:26 boris smidt <smidtboris1@gmail.com> wrote:

> No we just use a separate dependency.gradle file. I just wonder why we
> need an explicit _? Marker instead of just using _2.11 if you know your
> scala version is 2.11.12 . Also in the example of a project with the play
> plugin you can't combine it with your plugin so if it is defaulting then
> you don't need any special dependency configs. I.e. the plugin just
> specifies the version then and just uses the 2.11 2.12 2.13 as a marker. So
> other projects don't see the difference.
>
> On Fri, Jul 24, 2020, 22:22 Noam Asor <notifications@github.com> wrote:
>
>> Hi @borissmidt <https://github.com/borissmidt> :) Looks tidier now. Thx!
>> Helps a lot with review process! my question will be how come we need to do
>> replacement to actual _2.11, _2.12, _2.13 for the project to compile
>> correctly ? you should either have _? dependencies where you want the
>> plugin to expand it for you, based on the required cross compilation
>> versions or, you as the user, should provide those deps in correct
>> crosscompile configurations.
>>
>> Are you hinting that you are maintaining explicit 3rd party scala based
>> lib dependencies in build.gradle that are not aligned with default scala
>> version as defined by explicit scala-library dependency?
>>
>> As an example:
>>
>> project A {
>> apply plugin: 'com.github.prokod.gradle-crossbuild'
>>
>> dependencies {
>> compile some-lib_2.12:1.2.3
>> compile scala-lib:2.11.+
>> }
>> }
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/prokod/gradle-crossbuild-scala/pull/88#issuecomment-663718288>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ABNXZFRLTOEOT5UFA46WFCDR5HURHANCNFSM4PGAVSVQ>
>> .
>>
>

prokod commented:
This was a comment I have edited. But it seems you would like to get rid of
question marks all together. Am I getting you correctly?

So you would like to explicitly provide one scala version with no question
marks and let the plug-in modify it where needed.

Am I correct? And then, for your original issues in the specific project/s
you have tried original release/0.11… on with no success are you using
question marks anywhere in those?



On Sat, Jul 25, 2020, 8:12 PM Boris Smidt <notifications@github.com> wrote:

> A better question is:
>
> Would there be any case where you would want to have a _2.12 dependency in
> a _2.13 build?
>
> On Sat, Jul 25, 2020, 09:26 boris smidt <smidtboris1@gmail.com> wrote:
>
> > No we just use a separate dependency.gradle file. I just wonder why we
> > need an explicit _? Marker instead of just using _2.11 if you know your
> > scala version is 2.11.12 . Also in the example of a project with the play
> > plugin you can't combine it with your plugin so if it is defaulting then
> > you don't need any special dependency configs. I.e. the plugin just
> > specifies the version then and just uses the 2.11 2.12 2.13 as a marker.
> So
> > other projects don't see the difference.
> >
> > On Fri, Jul 24, 2020, 22:22 Noam Asor <notifications@github.com> wrote:
> >
> >> Hi @borissmidt <https://github.com/borissmidt> :) Looks tidier now.
> Thx!
> >> Helps a lot with review process! my question will be how come we need
> to do
> >> replacement to actual _2.11, _2.12, _2.13 for the project to compile
> >> correctly ? you should either have _? dependencies where you want the
> >> plugin to expand it for you, based on the required cross compilation
> >> versions or, you as the user, should provide those deps in correct
> >> crosscompile configurations.
> >>
> >> Are you hinting that you are maintaining explicit 3rd party scala based
> >> lib dependencies in build.gradle that are not aligned with default scala
> >> version as defined by explicit scala-library dependency?
> >>
> >> As an example:
> >>
> >> project A {
> >> apply plugin: 'com.github.prokod.gradle-crossbuild'
> >>
> >> dependencies {
> >> compile some-lib_2.12:1.2.3
> >> compile scala-lib:2.11.+
> >> }
> >> }
> >>
> >> —
> >> You are receiving this because you were mentioned.
> >> Reply to this email directly, view it on GitHub
> >> <
> https://github.com/prokod/gradle-crossbuild-scala/pull/88#issuecomment-663718288
> >,
> >> or unsubscribe
> >> <
> https://github.com/notifications/unsubscribe-auth/ABNXZFRLTOEOT5UFA46WFCDR5HURHANCNFSM4PGAVSVQ
> >
> >> .
> >>
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/prokod/gradle-crossbuild-scala/pull/88#issuecomment-663884734>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGWGZR225SQ4VJXET4ZF6TR5MOB5ANCNFSM4PGAVSVQ>
> .
>

borissmidt commented:
The questionmark should be best kept for legacy purposes and for archive
naming.

Yes the original 0.11 version did a better job that you could have a
dependency from a non crossbuild project to a crossbuild project and that
you get the dependencies of the correct version.

There was a bug then that you had to have a scala lang lib dependency in
all projects. I just used an allproject definintion to solve this.

So overall the plugin is getting better and i'm happy to see that happen!

On Sat, Jul 25, 2020, 22:17 Noam Asor <notifications@github.com> wrote:

> This was a comment I have edited. But it seems you would like to get rid of
> question marks all together. Am I getting you correctly?
>
> So you would like to explicitly provide one scala version with no question
> marks and let the plug-in modify it where needed.
>
> Am I correct? And then, for your original issues in the specific project/s
> you have tried original release/0.11… on with no success are you using
> question marks anywhere in those?
>
>
>
> On Sat, Jul 25, 2020, 8:12 PM Boris Smidt <notifications@github.com>
> wrote:
>
> > A better question is:
> >
> > Would there be any case where you would want to have a _2.12 dependency
> in
> > a _2.13 build?
> >
> > On Sat, Jul 25, 2020, 09:26 boris smidt <smidtboris1@gmail.com> wrote:
> >
> > > No we just use a separate dependency.gradle file. I just wonder why we
> > > need an explicit _? Marker instead of just using _2.11 if you know your
> > > scala version is 2.11.12 . Also in the example of a project with the
> play
> > > plugin you can't combine it with your plugin so if it is defaulting
> then
> > > you don't need any special dependency configs. I.e. the plugin just
> > > specifies the version then and just uses the 2.11 2.12 2.13 as a
> marker.
> > So
> > > other projects don't see the difference.
> > >
> > > On Fri, Jul 24, 2020, 22:22 Noam Asor <notifications@github.com>
> wrote:
> > >
> > >> Hi @borissmidt <https://github.com/borissmidt> :) Looks tidier now.
> > Thx!
> > >> Helps a lot with review process! my question will be how come we need
> > to do
> > >> replacement to actual _2.11, _2.12, _2.13 for the project to compile
> > >> correctly ? you should either have _? dependencies where you want the
> > >> plugin to expand it for you, based on the required cross compilation
> > >> versions or, you as the user, should provide those deps in correct
> > >> crosscompile configurations.
> > >>
> > >> Are you hinting that you are maintaining explicit 3rd party scala
> based
> > >> lib dependencies in build.gradle that are not aligned with default
> scala
> > >> version as defined by explicit scala-library dependency?
> > >>
> > >> As an example:
> > >>
> > >> project A {
> > >> apply plugin: 'com.github.prokod.gradle-crossbuild'
> > >>
> > >> dependencies {
> > >> compile some-lib_2.12:1.2.3
> > >> compile scala-lib:2.11.+
> > >> }
> > >> }
> > >>
> > >> —
> > >> You are receiving this because you were mentioned.
> > >> Reply to this email directly, view it on GitHub
> > >> <
> >
> https://github.com/prokod/gradle-crossbuild-scala/pull/88#issuecomment-663718288
> > >,
> > >> or unsubscribe
> > >> <
> >
> https://github.com/notifications/unsubscribe-auth/ABNXZFRLTOEOT5UFA46WFCDR5HURHANCNFSM4PGAVSVQ
> > >
> > >> .
> > >>
> > >
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/prokod/gradle-crossbuild-scala/pull/88#issuecomment-663884734
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAGWGZR225SQ4VJXET4ZF6TR5MOB5ANCNFSM4PGAVSVQ
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/prokod/gradle-crossbuild-scala/pull/88#issuecomment-663901648>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABNXZFVUFHJ4D23DYZU7FLLR5M4T3ANCNFSM4PGAVSVQ>
> .
>

borissmidt commented:
It would still be nice to skip the _? part since we use a depdencies.gradle file in quite a lot of projects where you define your dependencies and then use it in the subprojects. But then some are cross compiled and other are not and then you have to look out that you don't accidental have a cross compiled dependency.

prokod commented:
@borissmidt please see #80 it's about that and I plan this for next version. I preferred to have it in a separate release.

prokod commented:
superseded by #93
